providers/rpc/rpc.go

```diff
@@ -0,0 +1,351 @@
+	// CustomRequestPayload must be in json.
+	CustomRequestPayload []byte
+	// DotPath is the path to where the bmclib RequestPayload{} will be embedded. For example: object.data.body
+	DotPath string
+}
+
+// New returns a new Config containing all the defaults for the rpc provider.
+func New(consumerURL string, host string, secrets Secrets) *Config {
+	// defaults
+	c := &Config{
+		Host:        host,
+		ConsumerURL: consumerURL,
+		Logger:      logr.Discard(),
+		Opts: Opts{
+			Request: RequestOpts{
+				Client:          http.DefaultClient,
+				HTTPContentType: contentType,
+				HTTPMethod:      http.MethodPost,
+				TimestampFormat: time.RFC3339,
+				TimestampHeader: timestampHeader,
+			},
+			Signature: SignatureOpts{
+				HeaderName:             signatureHeader,
+				IncludedPayloadHeaders: []string{},
+			},
+			HMAC: HMACOpts{
+				Hashes:  map[Algorithm][]hash.Hash{},
+				Secrets: secrets,
+			},
+			Experimental: Experimental{},
+		},
+	}
+
+	if len(secrets) > 0 {
+		c.Opts.HMAC.Hashes = CreateHashes(secrets)
+	}
+
+	return c
+}
+
+// Name returns the name of this rpc provider.
+// Implements bmc.Provider interface
+func (c *Config) Name() string {
+	return ProviderName
+}
+
+// Open a connection to the rpc consumer.
+// For the rpc provider, Open means validating the Config and
+// that communication with the rpc consumer can be established.
+func (c *Config) Open(ctx context.Context) error {
+	// 1. validate consumerURL is a properly formatted URL.
+	// 2. validate that we can communicate with the rpc consumer.
+	u, err := url.Parse(c.ConsumerURL)
+	if err != nil {
+		return err
+	}
+	c.listenerURL = u
+	testReq, err := http.NewRequestWithContext(ctx, c.Opts.Request.HTTPMethod, c.listenerURL.String(), nil)
+	if err != nil {
+		return err
+	}
+	// test that we can communicate with the rpc consumer.
+	// and that it responses with the spec contract (Response{}).
+	if _, err := c.Opts.Request.Client.Do(testReq); err != nil { //nolint:bodyclose // not reading the body
+		return err
+	}
+
+	return nil
+}
+
+// Close a connection to the rpc consumer.
+func (c *Config) Close(_ context.Context) (err error) {
+	return nil
+}
+
+// BootDeviceSet sends a next boot device rpc notification.
+func (c *Config) BootDeviceSet(ctx context.Context, bootDevice string, setPersistent, efiBoot bool) (ok bool, err error) {
+	p := RequestPayload{
+		ID:     int64(time.Now().UnixNano()),
+		Host:   c.Host,
+		Method: BootDeviceMethod,
+		Params: BootDeviceParams{
+			Device:     bootDevice,
+			Persistent: setPersistent,
+			EFIBoot:    efiBoot,
+		},
+	}
+	rp, err := c.process(ctx, p)
+	if err != nil {
+		return false, err
+	}
+	if rp.Error != nil {
+		return false, fmt.Errorf("error from rpc consumer: %v", rp.Error)
+	}
+
+	return true, nil
+}
+
+// PowerSet sets the power state of a BMC machine.
+func (c *Config) PowerSet(ctx context.Context, state string) (ok bool, err error) {
+	switch strings.ToLower(state) {
+	case "on", "off", "cycle":
```

Comment:
I assume we don't have a concrete type for the states (enum style)?

Reply:
correct. we do not.